src/qt3utils/applications/qt3ple/main.py

```diff
@@ -0,0 +1,503 @@
+            self.root.quit()
+            self.root.destroy()
+        except Exception as e:
+            logger.debug(e)
+            pass
+
+    def scan_thread_function(self,
+                             vstart: float,
+                             vend: float,
+                             step_size: float,
+                             n_sample_size: int,
+                             n_scans: int,
+                             scan_mode: str,
+                             batch_size: int) -> None:
+        """
+        Function to be called in background thread
+        Scans the voltage from vstart to vend in increments of step_size
+        for a total of n_sample_size data points n_scans number of times.
+        If the scan_mode is "Batches", it will aggregate (average) some
+        batch_size number of points in between data points instead.
+        """
+        self.application_controller.set_scan_mode(scan_mode)
+        self.application_controller.discrete_batch_size = batch_size
+        self.application_controller.set_scan_range(vstart, vend)
+        self.application_controller.step_size = step_size
+        self.application_controller.tmax = n_scans
+        try:
+            self.application_controller.reset()  # clears the data
+            self.application_controller.start()  # starts the DAQ
+            self.application_controller.set_to_start()  # moves the stage to starting voltage
+
+            while self.application_controller.still_scanning():
+                self.application_controller.scan_wavelengths()
+                self.view.scan_view.update_image_and_plot(self.application_controller)
+                self.view.canvas.draw()
+
+            self.application_controller.stop()
+
+        except nidaqmx.errors.DaqError as e:
+            logger.info(e)
+            logger.info(
+                'Check for other applications using resources. If not, you may need to restart the application.')
+
+        self.enable_buttons()
+
+    def save_scan(self, event = None):
+        myformats = [('Pickle', '*.pkl')]
+        afile = tk.filedialog.asksaveasfilename(filetypes=myformats, defaultextension='.pkl')
+        logger.info(afile)
+        file_type = afile.split('.')[-1]
+        if afile is None or afile == '':
+            return # selection was canceled.
+        data = {}
+        data["Data"] = {}
+        if self.application_controller is not None:
+            for ii, scan_data in enumerate(self.application_controller.outputs):
+                data["Data"][f"Scan{ii}"] = {}
+                for reader in self.application_controller.readers:
+                    data["Data"][f"Scan{ii}"][reader] = scan_data[reader]
+        data["Metadata"] = []
+        if self.meta_configs is not None:
+            for meta_config in self.meta_configs:
+                data["Metadata"].append(meta_config)
+        if self.app_meta_data is not None:
+            data["ApplicationController"] = self.app_meta_data
```

Comment:
Can you implement npz and h5py, same as qt3scan options as well? Pickles are not the safest.

Reply:
It's a complicated dictionary that I was having trouble saving properly as a npz or h5py file. But I'll try again.